File: app/src/test/java/com/smlnskgmail/jaman/remotetemperaturecontrol/datavalidation/DataValidationTest.java
package com.smlnskgmail.jaman.remotetemperaturecontrol.datavalidation;

import com.smlnskgmail.jaman.remotetemperaturecontrol.monitor.connection.DataValidation;
import com.smlnskgmail.jaman.remotetemperaturecontrol.monitor.entities.signaltype.SignalType;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class DataValidationTest {

    private static final String DEFAULT_CLEAN_RESULT = "25.5";

    @Test
    public void validateTemperatureParse() {
        checkValidation(SignalType.Temperature);
    }

    @Test
    public void validateTemperatureMinimumParse() {
        checkValidation(SignalType.TemperatureMinimum);
    }

    @Test
    public void validateTemperatureMaximumParse() {
        checkValidation(SignalType.TemperatureMaximum);
    }

    @Test
    public void validateHumidityParse() {
        checkValidation(SignalType.Humidity);
    }

    @Test
    public void validateHumidityMinimumParse() {
        checkValidation(SignalType.HumidityMinimum);
    }

    @Test
    public void validateHumidityMaximumParse() {
        checkValidation(SignalType.HumidityMaximum);
    }

    @Test
    public void validateResetRequired() {
        checkValidation(SignalType.Reset, "");
    }

    private void checkValidation(SignalType signalType) {
        checkValidation(signalType, DEFAULT_CLEAN_RESULT);
    }

    private void checkValidation(SignalType signalType, String cleanResult) {
        String rawData = rawData(signalType, cleanResult);

        SignalType parsedSignalType = DataValidation.INSTANCE.signalType(rawData);
        assertEquals(parsedSignalType, signalType);

        String  parsedData = DataValidation.INSTANCE.data(rawData);
        assertEquals(parsedData, cleanResult);
    }

    private String rawData(SignalType signalType, String cleanResult) {
        return String.format("%s%s", signalType.getSignal(), cleanResult);
    }

}

File: app/src/test/java/com/smlnskgmail/jaman/remotetemperaturecontrol/datavalidation/DataValidationTest.kt
package com.smlnskgmail.jaman.remotetemperaturecontrol.datavalidation

import com.smlnskgmail.jaman.remotetemperaturecontrol.monitor.connection.DataValidation
import com.smlnskgmail.jaman.remotetemperaturecontrol.monitor.entities.signaltype.SignalType
import org.junit.Assert.assertEquals
import org.junit.Test

class DataValidationTest {

    @Test
    fun validateTemperatureParse() {
        checkValidation(SignalType.Temperature)
    }

    @Test
    fun validateTemperatureMinimumParse() {
        checkValidation(SignalType.TemperatureMinimum)
    }

    @Test
    fun validateTemperatureMaximumParse() {
        checkValidation(SignalType.TemperatureMaximum)
    }

    @Test
    fun validateHumidityParse() {
        checkValidation(SignalType.Humidity)
    }

    @Test
    fun validateHumidityMinimumParse() {
        checkValidation(SignalType.HumidityMinimum)
    }

    @Test
    fun validateHumidityMaximumParse() {
        checkValidation(SignalType.HumidityMaximum)
    }

    @Test
    fun validateResetRequired() {
        checkValidation(SignalType.Reset, "")
    }

    private fun checkValidation(signalType: SignalType, cleanResult: String = DEFAULT_CLEAN_RESULT) {
        val rawData = rawData(SignalType.signalOf(signalType), cleanResult)

        val parsedSignalType = DataValidation.signalType(rawData)
        assertEquals(parsedSignalType, signalType)

        val parsedData = DataValidation.data(rawData)
        assertEquals(parsedData, cleanResult)
    }

    private fun rawData(signalType: String, cleanResult: String): String {
        return String.format("%s%s", signalType, cleanResult)
    }

    companion object {

        private const val DEFAULT_CLEAN_RESULT = "25.5"

    }

}
